### DISCUSSION TITLE
👋 Welcome!

### DISCUSSION THREAD

**Discussion by pradeeban** [healthyinc/bio-block#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Chinmayee-cd** [comment]
Hello everyone!

My name is Chinmayee and I am an undergrad Computer Science third-year student. I am really interested in the Bio Block project. 

I have recently done a couple of projects on blockchain using Solidity and deployed them to Sepolia test network. Although I have not worked on the dual chain structure yet, I am keen to learn all the prerequisites and start implementation at the earliest. 

I would like to be part of this project as it aligns with my interests and skills, and I'm eager to bring value to the project while learning and growing through the journey!

Thank you.

**Reply by pradeeban** [reply]
Thanks, @Chinmayee-cd, for your interest. Nice to know you have relevant experience. Wishing you success with the GSoC 2025.

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Comment by karthiksathishjeemain** [comment]
Hello everyone, 
I am Karthik Sathish, a junior year student from IIT Roorkee, India. I have a good experience building applications on top of Ethereum blockchain. 

I am very interested in contributing to this project and would love to get started. However, I noticed that the repository currently contains only the README file and has not been initialized with any other code.

Could you please guide me on how to proceed? I am eager to resolve issues listed in the repository and contribute effectively. Any pointers on the next steps would be greatly appreciated.
Thank you.

**Reply by pradeeban** [reply]
Hi @karthiksathishjeemain,
Assuming you are a GSoC applicant: We are a GSoC organization with projects consisting of an equal mix of research and development. Some of our projects are new (including this one), and are marked as "New Project," while others are well-developed. 

If you would like an adventure of developing a new project with us, you should choose this or any other project that are labeled "New Project" (such as [Diomede](https://github.com/KathiraveluLab/Diomede)). There are certain benefits to new projects, including greater freedom and influence in designing it from scratch with us. Then, there are also obvious challenges in dealing with unknown waters. Some students prefer new projects, while others prefer an already-established codebase. We offer both as "Alaska", a largely a university system-based organization.

If you would prefer a more well-developed project, I recommend you checking other project ideas listed in https://github.com/uaanchorage/GSoC: [CoastlineExtraction](https://github.com/fwitmer/CoastlineExtraction), [xlns](https://github.com/xlnsresearch/xlns), [Metagenome](https://github.com/akdasUAF/Metagenome), [L4SBOA](https://github.com/KathiraveluLab/L4SBOA), [beehive](https://github.com/kathiraveluLab/beehive), [concore](https://github.com/ControlCore-Project/concore), and [AWANTA](https://github.com/KathiraveluLab/AWANTA) instead.

**Reply by karthiksathishjeemain** [reply]
Thanks for the response, @pradeeban. Yes, I am a GSoC applicant. I have an idea on how to proceed with this project. Would you prefer that I write it down here, or should I submit it as a formal proposal via email?

**Reply by pradeeban** [reply]
You could write down here. During later stages (once/if Alaska is accepted to GSoC 2025), you could upload the draft proposal to the GSoC site and simply request us for a review by commenting here.

**Reply by karthiksathishjeemain** [reply]
Sure sir

**Comment by karthiksathishjeemain** [comment]
@pradeeban Here is the proposed workflow for the project:

Firstly, a user accesses the decentralized web interface built with React, which is integrated with their blockchain wallet (such as MetaMask). Instead of scraping external data, the system empowers users to directly submit their biological or medical data through intuitive React components. Upon data upload, a hashing function computes a unique fingerprint for the submission. This hash is then recorded on the blockchain, ensuring data integrity and establishing proof of ownership.

Once the data is verified, it is processed for efficient storage. If necessary, the data is divided into manageable chunks, each associated with its own metadata. To maintain scalability and efficiency, the actual files are stored off-chain using a decentralized storage solution like IPFS. The blockchain only stores the pointers (hashes and IPFS addresses) linking to the data, thereby creating a secure, tamper-proof registry without overloading the blockchain with large files.

With the data securely stored, the dual-chain architecture becomes active. One chain—the Data Chain—maintains a ledger of data submissions, ownership records, and metadata. The other—the Payment Chain—manages all financial transactions related to accessing the data. Each submission is tokenized, providing data owners with a transparent and auditable record of their contributions.

When a researcher or data consumer initiates a query via the React-based interface, the request is first validated by a smart contract. This contract checks the user’s credentials and ensures they possess sufficient funds for data access. The query then triggers a search mechanism on the Data Chain that retrieves the relevant data pointers. Similar to methods like maximum marginal relevance (MMR) or BM25 used for document matching, the system identifies and returns the most pertinent data pointers (an approach I successfully implemented during my recent internship with positive results).


After identifying the correct data pointers, the smart contract on the Payment Chain securely processes the token transfer. The payment is automatically allocated to the data owner’s wallet, ensuring immediate compensation. Once the transaction is complete, the data pointers are relayed back to the consumer, who can then use the pointers to retrieve the full data from the decentralized storage system.

Additionally, the platform supports alternative modes of interaction. For instance, if a user prefers using voice commands, the system integrates with speech-to-text models such as Assembly AI or OpenAI’s Whisper. The audio input is transcribed into text and processed just like a standard text query. The React frontend then displays the human-readable response generated from the retrieved data, ensuring a consistent user experience across input modalities.

The end result is a robust agent that leverages blockchain’s security and decentralization to guarantee data ownership and seamless payment processing. Data consumers receive precise, relevant information directly from the decentralized data repository, while data owners are automatically compensated for their contributions.

The tech stack supporting this workflow includes:

Frontend: React for a dynamic, user-friendly web interface.
Smart Contracts: Solidity on the Ethereum blockchain.
Backend Integration: Python with web3.py for blockchain communication.
Off-Chain Storage: IPFS for scalable, decentralized data storage.
Voice Integration: Speech-to-text models such as Assembly AI or OpenAI’s Whisper.

Please let me know if there is any part which requires further detailing. 
Also, I highly appreciate any suggestions that can enhance my above workflow.

**Reply by karthiksathishjeemain** [reply]
@pradeeban does this workflow requires any further enhancement?

**Reply by pradeeban** [reply]
@karthiksathishjeemain It looks good to me. I like your choices of React, IPFS, and Solidity. The mentors will share additional feedback if they have any. Good work!

**Reply by Chali-healthy** [reply]
Hello Karthik! We look forward to looking at your final submission to the project. Please refer to additional comments that @pradeeban made regarding additional questions to consider in your final submission.

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Reply by karthiksathishjeemain** [reply]
Sure @Chali-healthy @pradeeban , I have started writing the proposal. I will share it with you before submitting it to the GSOC portal. Thanks for the remainder

**Reply by Chali-healthy** [reply]
Thanks Karthik!

**Reply by karthiksathishjeemain** [reply]
@Chali-healthy sir, most part of the proposal is done. I have a small doubt : Is it okay to include a work-flow diagram in my proposal?
I feel that including that will make the mentors understand my proposal much better.

**Reply by pradeeban** [reply]
Totally! Workflows are a great way to showcase your thought process.

**Reply by karthiksathishjeemain** [reply]
Updates:
I have created workflows for most of the processes and a few is remaining. I will write the timeline section and submit my proposal for review by April 3rd IST.

**Reply by pradeeban** [reply]
Sounds good. Let us know once you have uploaded!

**Reply by karthiksathishjeemain** [reply]
@pradeeban @Chali-healthy sir, I have sent you the proposal draft for review. Please let me know if it needs any changes.

**Reply by pradeeban** [reply]
@karthiksathishjeemain Neat! I am going to reply shortly. Make sure to also submit on GSoC Portal as soon as possible. You can keep re-submitting until the deadline in the GSoC Portal.

**Reply by karthiksathishjeemain** [reply]
Yep, I submitted on the GSOC portal as well.

**Reply by pradeeban** [reply]
Yes, saw it now. Also replied to your email.

**Reply by karthiksathishjeemain** [reply]
@pradeeban sir, as the proposal submission period has ended now, can we start brainstorming on our Idea by implementing small examples?

**Reply by pradeeban** [reply]
@karthiksathishjeemain go for it! Although just be aware that as per the GSoC policies, the developments made after the GSoC application due date shouldn't impact the decision process in accepting to GSoC.

**Reply by karthiksathishjeemain** [reply]
Sure sir.

**Comment by prab0007al** [comment]
Hello everyone,
I’m Prabal Thool, a B.Tech Computer Science Engineering student at IIT Mandi. I have a strong interest in blockchain technology, cybersecurity, and decentralized applications. My technical skills include proficiency in C/C++, Python, and JavaScript, and I am currently expanding my knowledge in Ethereum-based blockchain development. I’m particularly excited about Bio-Block’s vision of creating a decentralized, secure, and scalable data repository. I look forward to learning from this community, contributing to discussions, and collaborating on building this innovative platform.

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Comment by zeroxraghav** [comment]
Hello everyone!

My name is Raghav. I am an undergrad ECE student from NIT Surat. 

I have worked actively in the blockchain and DeFi space for several months. I have hands-on experience in building Solidity smart contracts. I'm fascinated by the workings of DeFi protocols, constantly learning new terminologies and exploring core blockchain concepts I haven’t encountered before.

I am highly interested in contributing to the Bio-Block project. I have several ideas to strengthen the platform, such as using Layer-2 solutions like zkSync for gas-efficient transactions, utilizing decentralized storage solutions like Arweave/IPFS, and tokenizing biodata using ERC-20 tokens or NFTs to enable ownership and monetization.

While my primary experience is with Foundry, I am open to learning new frameworks and technologies to contribute effectively. I’m excited to collaborate with this community and work towards making this project successful.

Looking forward to building together! 
Thank you.

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Comment by sarthakeash** [comment]
Hey! I am Sarthak Arora. I graduated CS from BITS Pilani last year. I am currently working as a software developer in a US based SasS. Open source is new to me and I always wanted to get my hands into it, hence applying to Gsoc this year. The Bio-block project specifically seems very interesting to me. I have several ideas including a dual chain structure, using ipfs/ file coin for effective data distribution/encryption, using synthetic pipelines etc. 
I am happy to discuss more about these, looking forward!

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Comment by OneFriendlyCoder** [comment]
Hello everyone,
I am Abhyanand Sharma, currently pursuing my masters in Computer Science at Indian Institute of Technology Bombay, with a strong background in blockchain and full-stack development. My hands-on experience includes designing and deploying Hyperledger Fabric networks, building real-time Ethereum NFT trackers, and developing smart contracts and dynamic applications using React, Python, and TypeScript. I have also developed a P2P crypto simulator, reflecting my intensive knowledge of both layer 1 and layer 2 blockchain protocols, and I possess a deep understanding of blockchain transactions and payment methods. I am excited by this project’s vision of creating a secure, decentralized bio-block that empowers data owners, and I look forward to contributing my skills to advance this innovative initiative.

**Reply by Chali-healthy** [reply]
Hello Abhyanand! Thank you for your email follow up. We look forward to looking at your final submission to the project. Please refer to additional comments that @pradeeban made regarding additional questions to consider in your final submission.

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Comment by imshrishk** [comment]
@pradeeban This is what I have come up with after a bit of thought and believe it to be a good enough solution for the project. Please suggest if there is anything missing or if any part of the workflow is incorrect. I would also love to provide more details on any part if you want. I have formatted it for better visibility. I was also wondering about the code challenge part where I have to share my github repo. I have made projects that are private for an organisation and hence cannot share it, is there any other alternative to show my skills in the sector?

# Workflow Proposal

## User Authentication & Data Submission:
- A user accesses the decentralized web interface built with React, which integrates with their blockchain wallet (e.g., MetaMask, Trust Wallet, etc.) we will have to see what suits our need the most).  
- Users submit their biological or medical data directly through the forms or some similar process in our frontend.  
- A hashing function computes a unique fingerprint for the submission as the data is uploaded and is stored on blockchain, ensuring data integrity, immutability, and proof of ownership.  

## Efficient Storage & Data Management:
- The data undergoes validation and segmentation into manageable chunks (if the data is large), each with associated metadata.  
- For scalability, IPFS (InterPlanetary File System) can be used to store the actual data files off-chain.  
- The blockchain maintains only pointers (hashes and IPFS addresses), making it secure and avoiding blockchain overload.  

## Dual-Chain Architecture Activation:
- **Data Chain**: Maintains records of data submissions, ownership, and metadata.  
- **Payment Chain**: Manages financial transactions for data access.  
- Each data submission is tokenized, creating an auditable and transparent record for data owners.  

## Data Query & Access:
- A researcher can submit their queries to get the data using a form-like system similar to the one used for inputting the data.  
- The request is validated by a smart contract, verifying credentials and available funds before processing further.  
- The system searches the Data Chain for relevant records using a retrieval approach providing us the data pointers. (This can be a normal approach or optionally a RAG).  

## Payment:
- The smart contract on the Payment Chain securely processes token transfers.  
- The payment is automatically allocated to the data owner’s wallet to enable instant compensation.  
- Once the transaction is complete, the data pointers are provided to the consumer, who retrieves the data from IPFS.  

## Security & Governance:
- **Homomorphic Encryption** ensures privacy while allowing computations on encrypted data.  
- **Zero-Knowledge Proofs (ZKP)** provide an extra layer of authentication.  
- A **Decentralized Autonomous Organization (DAO)** governs platform upgrades and pricing mechanisms. (This might be too far-fetched but is still an idea).  

---

# Proposed Technology Stack  

## Blockchain & Storage:
- **Ethereum Blockchain** (Primary ledger for metadata and transactions)  
- **IPFS** (Decentralized storage for raw biological data)  

## Smart Contracts & Payments:
- **Solidity** (Smart contract development)  
- **OpenZeppelin** (Security standards for smart contracts)  
- **ERC-20 & ERC-721 Tokens** (Tokenized payments and data ownership tracking)  

## Backend & API Layer:
- **Python (Flask/FastAPI)** (For backend API development)  
- **Web3.py** (Interfacing with Ethereum blockchain)  
- **IPFS-Py** (Interacting with IPFS for storage)  
- **PostgreSQL** (Metadata storage for quick lookups)  

## Frontend:
- **React.js** (User Interface)  
- **Ethers.js** (Blockchain interaction in frontend)  

## Security & Privacy:
- **Homomorphic Encryption** (Data encryption while allowing computations)  
- **Zero-Knowledge Proofs (ZKP)** (Privacy-preserving verification)  
- **OAuth2 & JWT** (User authentication, sign-in & authorization)  

---

# Optional Enhancements  

## RAG-Powered Search Mechanism:
- Implement a vector search using an open-source model like **Gemma, Llama,** or any similar models for improved data retrieval.  
- Utilize **FAISS** or **ChromaDB** for optimized query performance. (Personally, I would prefer ChromaDB due to past experience with it).  

## Alternative Interactions (Speech-to-Text Integration):
- Allow users to interact using voice commands via **ElevenLabs S4TS** or **Kaldi**.  
- Convert voice inputs into text for further modification by the user in case our model got it wrong, allowing corrections before starting the query.  

## Multi-Blockchain Interoperability:
- Expand support to **Binance Smart Chain, Solana, or Avalanche**, etc.  
- Enable **cross-chain transactions** for enhanced accessibility.  

## Data Monetization Models:
- Implement **subscription-based access** for research institutions.  
- Enable **data auctions** for high-value genomic datasets. (This might be possible for a new discovery of a species or anything similar which is incredibly rare and mostly won’t be used, but just an idea to throw around).  

## Edge Computing for Data Processing:
- Enable lightweight **on-device processing** for privacy-sensitive data.  
- Utilize **WASM-based smart contracts** for decentralized computation.

**Reply by pradeeban** [reply]
You could share code segments, that you think could help, in your proposal (that you submit at GSoC portal). Make sure you have permissions from your organization to share the code segments you are sharing.

**Reply by imshrishk** [reply]
Will it put me in a disadvantage compared to others, if yes, should I make a small 2-3 day project on these topics and submit that there? Also, I would appreciate comments on the above workflow to make it better. Thanks a lot!

**Reply by pradeeban** [reply]
The workflow looks good to me. There are many alternative implementation options since the project is still in its early phase, with portions of research and implementation. You could always create dummy projects in your repository to include links in your proposal, if you have time and interest.

**Reply by Chali-healthy** [reply]
Hello! We look forward to looking at your final submission to the project. Please refer to additional comments that @pradeeban made regarding additional questions to consider in your final submission.

**Reply by pradeeban** [reply]
Pls make sure to submit the proposal on the GSoC Portal. Ideally, immediately. You can continue to update until April 8th. No updates made after GSoC deadline will be considered. Make sure to answer the questions raised in https://github.com/bio-block/healthy/discussions/5 in your proposal.

**Reply by imshrishk** [reply]
@pradeeban @Chali-healthy I have added my proposal to the GSoC portal. I know it is pretty late but I tried my best to make a sample project and it took a little more time than expected. I have also tried to answer all questions raised in #5 as well I could have. Is it still possible to get a review before the time period ends to make any changes if needed?

**Reply by pradeeban** [reply]
Hi @imshrishk, we have 76 proposals (and counting). So, please be patient with us. We will get back if we have additional comments or questions. Otherwise, it is more likely you will hear directly from Google, than us.

**Reply by imshrishk** [reply]
Thanks a lot!

**Comment by prathameshParadkar** [comment]


**Reply by pradeeban** [reply]
1. It should be. The platform should be able to support that.
2. Yes, please. But remember that the application period also has started. So make sure to submit the application on time as well.

**Comment by OneFriendlyCoder** [comment]
Hi @pradeeban @Chali-healthy ,
I have submitted by GSOC 2025 proposal for the same.
Thank you

**Reply by pradeeban** [reply]
awesome, thanks, also received your email. Will go through it and reply if I have comments or questions.

**Comment by prathameshParadkar** [comment]
Hi @pradeeban @Chali-healthy,
I have mailed you the final draft of my proposal as well as submitted it on GSOC 2025.
Please let me know if any changes are required.
Thank you

**Comment by santhil-cyber** [comment]
@Chali-healthy  @karthiksathishjeemain  @pradeeban here is my rough proposed workflow for the Bio-Block Data Analytics dApp(want some feedback):

firstly, a user who has already purchased a dataset through bio-block's existing `purchaseDocument` flow accesses the analytics dashboard through the React frontend. their identity is verified using EIP-712 wallet signatures  the buyer signs a message with their MetaMask wallet, and the analytics service verifies this signature against the `DocumentPurchased` event logs on-chain. this means we dont need a separate authentication system; theres a direct cryptographic link between the analytics request and the on-chain purchase proof.

once authenticated, the frontend retrieves the encrypted dataset from IPFS using the stored CID and decrypts it client-side using the existing encryption key mechanism thats already built into bio-block. the decrypted data is then sent to a dedicated Analytics API a separate FastAPI microservice running alongside the existing python backend. i chose to keep it separate because the analytics dependencies (pandas, scipy, scikit-learn, statsmodels) are heavy and shouldnt bloat the core metadata service. this also means the analytics engine can be scaled independently, and if it crashes during a heavy computation, it doesnt take down the search, store, or anonymization endpoints that other users depend on.

the analytics engine itself implements the four levels of analysis described in the project brief. at the graphical and descriptive level, the engine uses pandas for data profiling (summary statistics, missing values, column distributions, data types) and matplotlib/plotly for generating interactive visualizations like distribution plots, correlation heatmaps, and time series charts. at the inferential level, scipy.stats and statsmodels handle hypothesis testing (t-tests, chi-square tests, ANOVA), confidence intervals, and correlation analysis (pearson and spearman). predictive analytics using scikit-learn (linear regression, basic classification) is planned as a stretch goal, since building reliable ML pipelines on arbitrary unsanitized health datasets requires careful feature engineering and validation that may exceed the GSoC timeline if rushed.

to protect against resource exhaustion, the API enforces strict input limits file uploads are capped at 50MB / 50k rows, heavier jobs like inferential analysis run asynchronously through a celery/redis task queue so the API doesnt hang, and per-wallet rate limiting prevents abuse. all inputs go through schema validation (column types, missing data thresholds) before any compute begins.

on the privacy side, the data has already been anonymized by bio-block's existing presidio and OCR pipeline before it was ever uploaded to IPFS, so the analytics layer is working with de-identified data. however, aggregated statistical outputs can still leak information for example, if a descriptive analysis returns stats for a cohort of only 3 people. to address this, i plan to add an optional differential privacy layer using OpenDP (the open-source library from Harvard and Microsoft). for descriptive stats like means and counts, calibrated Laplace noise is applied with a configurable epsilon (default ε=1.0). for inferential outputs like p-values and correlations where noise would destroy statistical validity, results from small cohorts (fewer than 30 samples) are flagged with a warning instead. for visualizations, data is binned and aggregated before rendering so individual data points arent exposed in scatter plots. the privacy budget is user-configurable per analysis session so researchers can tune the privacy-utility tradeoff based on their dataset's sensitivity.

once an analysis is complete, the results are serialized as a structured JSON report along with any generated visualizations, uploaded to IPFS through the existing pinata integration in the javascript backend, and then registered on-chain via a new `AnalyticsRegistry` smart contract. this contract links the result CID to the source dataset CID, creating an immutable audit trail anyone can verify that a specific analytics result was derived from a specific dataset. the contract is intentionally lightweight (just a registry, no on-chain computation) and will be deployed on sepolia testnet for the GSoC deliverable, designed to be L2-deployable (Polygon or Arbitrum) post-GSoC when theres real usage data to justify the migration.

the frontend dashboard, built with chart.js or d3, provides an interactive interface where users can select their purchased datasets, choose analysis types, configure parameters, and view results with properly rendered visualizations. completed analyses are saved and linked to their on-chain attestations for future reference.

ive been contributing to bio-block for the past few weeks and have built up a solid understanding of how the different components interact:
- fixed search endpoints querying wrong ChromaDB collection (#144)
- fixed duplicate function naming in store endpoint (#142)
- restored the anonymize_image fallback endpoint (#139)
- patched 6 security vulnerabilities in DocumentStorage.sol (#155, #152, #176)

this gave me a clear picture of the python backend's FastAPI structure, the smart contract architecture, the IPFS/pinata integration, and the chromadb vector store — all of which the analytics dApp needs to integrate with cleanly.

the tech stack supporting this workflow includes:

- **analytics engine**: python with FastAPI, pandas, scipy, statsmodels, scikit-learn, OpenDP
- **frontend**: react with chart.js/d3 for interactive dashboards
- **smart contracts**: solidity on ethereum (sepolia testnet, L2-ready)
- **authentication**: EIP-712 wallet signatures verified against on-chain purchase events
- **off-chain storage**: IPFS via pinata for analytics results
- **task processing**: celery + redis for async heavy compute jobs

please let me know if any part needs further detailing or if there are specific analytics use cases for health data that should be prioritized (cohort analysis, survival curves, epidemiological measures, etc). id also appreciate any thoughts on the differential privacy approach and whether the current epsilon defaults make sense for bio-block's typical dataset sizes.

looking forward to your feedback.

**Reply by karthiksathishjeemain** [reply]
Thanks for the detailed workflow for the Data Analytics project. I went through it. You explained it greatly in most of the places, especially the System Design(good work with that).
These are my feedback/questions for the proposed workflow:
1. Why just decrypt on frontend but analyze on backend? Since privacy is most important aspect when we are dealing with medical related dataset, can we analyse data on the frontend itself(along with decrypting it) or have a secure conclave for doing so? Sending the fully decrypted data to a server running elsewhere has a potential risk of getting the data leaked (if you look at the current architecture, we have ensured to keep the clients data(decrypted version) from hitting any external servers.
2. I was unable to clearly relate with the use case of OpenDP here (a bit more detailing would have been better).

Feel free to ask questions here. Great job with the workflow!!!

**Reply by santhil-cyber** [reply]
Thank you @karthiksathishjeemain for going through it and the kind words on the system design! both questions are really important, let me address them properly.

**on frontend-only analytics vs sending decrypted data to a backend:**

you're absolutely right, and honestly this is something i should have caught earlier. looking back at the existing architecture — the frontend encrypts files before they leave the browser, the backend only ever handles encrypted blobs, and decryption happens client-side. sending fully decrypted data to a separate analytics server would break that exact guarantee.

so here's my revised thinking after some research - a hybrid client-first approach:

for level 1 and 2 analytics (graphical + descriptive), we can do everything entirely in the browser. the JavaScript ecosystem actually has solid libraries for this:

- **Danfo.js** — basically pandas for JavaScript, built by the TensorFlow team. handles dataframes, summary stats, groupby, missing value analysis, all client-side
- **Plotly.js / Chart.js** — interactive charts and visualizations, already runs in the browser
- **simple-statistics** — mean, median, standard deviation, histograms, percentiles — all the descriptive stats you'd need

for level 3 (inferential), things like t-tests, chi-square, and correlation analysis are mathematically simple enough to run in the browser using `jStat` (a JavaScript statistical library). for more complex stuff like ANOVA or regression, we could use **Pyodide** (Python compiled to WebAssembly) — this literally runs pandas and scipy inside the browser sandbox. the data never leaves the user's machine.

for level 4 (predictive / stretch goal) — this is the only case where client-side gets genuinely limiting (training ML models in the browser is slow and memory-constrained). for this, we have two options:

1. keep it client-side using **TensorFlow.js** for basic models (linear regression, simple classification) — slower but fully private
2. offer an opt-in server mode with explicit user consent, where the data is sent over TLS to an ephemeral processing container that has no persistent storage. the data exists in memory only for the duration of the computation and is wiped after. this would be clearly labeled as "server-assisted analysis" so the user makes an informed choice

but honestly, if we nail levels 1-3 fully client-side within the GSoC timeline, that already satisfies the project requirement of "at least two levels of analysis" and keeps the privacy model completely intact.

**Reply by santhil-cyber** [reply]
**on OpenDP let me give a concrete example:**

the reason i brought up differential privacy is specifically for the outputs of analytics, not the raw data. here's the scenario:

say a researcher purchases a dataset of patient lab results and runs a descriptive analysis filtered by `age_group = "70-80" AND diagnosis = "diabetes"`. if there are only 3 patients matching that filter, the output stats (`mean blood sugar = 248`, `std dev = 12`) could potentially identify those individuals especially if someone already knows that *"there are 3 diabetic patients over 70 in this hospital."*

**OpenDP** (built by Harvard's Privacy Tools Project and Microsoft) solves this by adding carefully calibrated noise to aggregated outputs:
```
without DP:  mean_blood_sugar = 248.3  (exact — potentially identifying)
with DP:     mean_blood_sugar = 251.7  (noised — still useful, but not exact)
```

the noise is calibrated using an epsilon (ε) parameter:

- `ε = 0.1` → heavy noise, very private, less accurate
- `ε = 1.0` → balanced (the standard default)
- `ε = 10.0` → minimal noise, very accurate, less private

the key insight is that this applies to the analytics results shown to the user, not to the underlying data. so even in the client-side architecture, after Danfo.js computes the statistics in the browser, the DP layer adds noise before rendering the charts. this protects against statistical re-identification attacks on small cohorts.

**Reply by pradeeban** [reply]
Interesting insights, @santhil-cyber.

Looking forward to your proposal draft.

@karthiksathishjeemain @Chali-healthy

**Reply by santhil-cyber** [reply]
Thank you sir @pradeeban

**Reply by Chali-healthy** [reply]
Look forward to your proposal @santhil-cyber!

**Reply by santhil-cyber** [reply]
Thank you @Chali-healthy! Looking forward to your feedback

**Reply by karthiksathishjeemain** [reply]
@santhil-cyber, I just sent you some feedback regarding your current proposal over the email. Looking forward to your submission!!

**Reply by santhil-cyber** [reply]
Feedback taken!!🫡...Will implement it Thanks @karthiksathishjeemain

**Reply by santhil-cyber** [reply]
Hi @pradeeban @Chali-healthy @karthiksathishjeemain

Quick update I have submitted my proposal draft on the GSoC portal. It incorporates the feedback @karthiksathishjeemain shared over email.
I genuinely appreciate how welcoming and supportive you all have been throughout this process. It means a lot.

Happy to answer any questions or iterate further based on your feedback.

**Reply by pradeeban** [reply]
I went through your proposal in the GSoC Portal, @santhil-cyber.
Nicely written.

**Reply by santhil-cyber** [reply]
Glad to hear that, @pradeeban sir!! Open to any feedback. 🙌

**Reply by santhil-cyber** [reply]
Hey @pradeeban @Chali-healthy  @karthiksathishjeemain , realizing it might be time consuming for you to set up and run the backend POC (which I mentioned in proposal) locally, I put together a quick 1-minute demo video with a short demo data. It demonstrates exactly what I outlined in the proposal the API handling the EIP-712 security checks and returning the statistical analysis. 

You can check it out here: https://youtu.be/43gZq2gAIZI

Result: https://drive.google.com/drive/folders/1D2KMtV4WPPhI24QFNsq8f2VUit4_Jm0O?usp=sharing

**Comment by VincenzoImp** [comment]
Hi everyone, I'm Vincenzo Imperati, PhD student at Sapienza University of Rome. I'm interested in the Bio-Block Data Analytics dApp project for GSoC 2026.

I've already submitted several contributions: a bug fix for the `/search` endpoint (PR #161, merged), CI/CD fixes (PR #171, merged), input validation (PR #166), an ownership hijacking fix in `storeDocument` (PR #167), and test suites for PHI detection and IPFS upload (PRs #168, #169).

My background is in large-scale data analysis and graph modeling. I maintain BigBrotr, an OpenSats-funded Nostr relay observatory with 8 async Python services and PostgreSQL 16.

**Reply by pradeeban** [reply]
Impressive profile and contributions @VincenzoImp.

**Reply by Chali-healthy** [reply]
@VincenzoImp look forward to the proposal.

**Reply by karthiksathishjeemain** [reply]
Good work @VincenzoImp!! Looking forward to your proposal.

**Comment by SIDDHANTCOOKIE** [comment]
Hey @pradeeban @Chali-healthy @karthiksathishjeemain , I have just mailed my proposal draft with all information to you all and cc'd other mentors as well. Please review it and provide your valuable feedback on the proposed architecture so I can refine it before the final submission deadline. Thank you!

**Reply by pradeeban** [reply]
@SIDDHANTCOOKIE Received. Thanks.

**Reply by SIDDHANTCOOKIE** [reply]
Hey @pradeeban  @Chali-healthy @karthiksathishjeemain , 

Previous few days while working on the proposal ​I was thinking about the long-term working of Bio-Block, specifically regarding longitudinal data tracking in the context of the analytics engine. 

​The platform currently uses Wallet-based hashing for personal data to maintain anonymity. This is a great architecture for individual users,because the hash is tied to their personal wallet, our analytics backend can theoretically group their future uploads together for time-series charts without knowing their real identity. 

​However, a thought crossed my mind regarding institutional sellers:   
What happens if a clinic or hospital uploads data for 1,000 different patients? If the anonymization hash is derived from the clinic's wallet, wouldn't all 1,000 patients effectively be merged under the same anonymous identity? And if that clinic comes back next year to upload updated records for those exact same patients, how do we link the Year 1 and Year 2 records for individual patients without breaking anonymity? 

Is there a way we can show progressive tracking and link these longitudinal records over time, without accidentally breaking the anonymity of the patients?

**Reply by pradeeban** [reply]
@SIDDHANTCOOKIE Storing the hash of a PatientID (such as EMPI) should help, I believe. Sure, the uploader/owner will be the hospital. There are certain HIPAA/privacy concerns to consider. You can leave those as open questions. We came across the same issue with Niffler DICOM anonymizer several years ago. We stored the data anonymized. EMPI (the PatientID) itself can be considered HIPAA. So, we anonymized it as a hash.

**Reply by SIDDHANTCOOKIE** [reply]
@pradeeban,
Thanks for the insight! Yes, this is what crossed my mind when thinking about the long-term roadmap after the initial GSoC dApp completion, especially regarding how strict HIPAA compliance might bottleneck future features.
Your approach with the Niffler DICOM anonymizer is a great precedent. Storing the EMPI as a cryptographic hash is an elegant solution to maintain that longitudinal linkage without ever exposing the actual PatientID to the analytics engine.
As you suggested, I'll leave the deeper HIPAA compliance and management as open questions for the future roadmap :)

**Reply by pradeeban** [reply]
Yes, the EMPI (which was a UUID, the PatientID) itself is a PHI. The hash or a ciphertext of EMPI works better. Better than just using EMPI. But how comfortable we will be with loading a hash/ciphertext of EMPI that is unique to a patient: that we will see.

**Reply by SIDDHANTCOOKIE** [reply]
That is a very fair point, @pradeeban. Even with a hash or ciphertext, a persistent unique identifier creates a "linkability" risk over time that we have to be cautious about.
If the same hash follows a patient across every single dataset, we're essentially creating a pseudonymous shadow-identity that could theoretically be re-identified through side-channel data or frequency analysis.
Thanks for pointing this out! It’s definitely a complex privacy tradeoff and not an easy one to solve. We can have a good brainstorm on this further with everyone post-dApp development once we have the foundational analytics in place. Really appreciate your insights here!

**Reply by pradeeban** [reply]
Precisely. We still want a 1:1 mapping between the EMPI and its "hash"/cipher. That creates a potential/risk for deanonymization. We really did not think into that far (because this is more of a policy question than an implementation challenge).

**Reply by SIDDHANTCOOKIE** [reply]
Exactly! It definitely crosses over from a pure engineering challenge into a platform policy decision. It’s going to be a really exciting bridge to cross once the platform reaches that scale.  
Thanks for this insightful discussion, @pradeeban it gave me a lot of great context!!!

**Reply by SIDDHANTCOOKIE** [reply]
@pradeeban @Chali-healthy @karthiksathishjeemain,  I've submitted my proposal on the GSoC Portal as discussed, awaiting your feedback!

**Reply by pradeeban** [reply]
Awesome, good luck, @SIDDHANTCOOKIE
